### PR TITLE
podman kube play --replace should force removal of pods and containers

### PR DESCRIPTION
**Pull request by rhatdan**
Fixes: https://github.com/containers/podman/issues/20025

#### Does this PR introduce a user-facing change?

```release-note
None
```
[NO NEW TESTS NEEDED]

### CODE REVIEW

File: pkg/bindings/kube/types.go [hunk 1]
```diff
@@ -42,6 +42,8 @@ type PlayOptions struct {
 	LogDriver *string
 	// LogOptions for the container. For example: journald
 	LogOptions *[]string
+	// Replace - replace existing pods and containers
+	Replace *bool
```
Comment (vrothberg) [inline pkg/bindings/kube/types.go:46]: That screams for tests.  It seems `kube play --replace` is not tested on the remote clients or the tests are incomplete.

File: pkg/domain/infra/abi/play.go [hunk 2]
```diff
@@ -1511,7 +1511,7 @@ func (ic *ContainerEngine) PlayKubeDown(ctx context.Context, body io.Reader, opt
 		return nil, err
 	}
 
-	reports.RmReport, err = ic.PodRm(ctx, podNames, entities.PodRmOptions{Ignore: true})
+	reports.RmReport, err = ic.PodRm(ctx, podNames, entities.PodRmOptions{Ignore: true, Force: true})
```
Comment (vrothberg) [inline pkg/domain/infra/abi/play.go:1514]: That only sets `pod rm --force` which will stop running containers instead of erroring out.

**Review by vrothberg — COMMENTED** [review]
Please split the doc changes into a separate commit.

Looking purely at the changes, I don't think it's sufficient to fix https://github.com/containers/podman/issues/20025?  Please add a regression test to make sure it does.  The problem in #20025 is that "external" containers are not being removed/replaced, so we could probably do a `buildah from --name=$expected-name` and use that name in a kube YAML to cause a name collision.

File: pkg/domain/infra/abi/play.go (hunk 2 above)
Comment (rhatdan) [reply pkg/domain/infra/abi/play.go:1514]: I believe it will also push the force to remove the containers as well.

Comment (rhatdan) [reply pkg/domain/infra/abi/play.go:1514]: But if I can get the buildah test to work, then we can actually see if it works.

Comment (vrothberg) [reply pkg/domain/infra/abi/play.go:1514]: I did not test it yet but my assumption is that this is not enough for the Buildah container.  Imagine only one container exists ... the pod won't exist, so there's is nothing to remove.  I'd expect the container to not be removed then.  But I may be off.

Comment (rhatdan) [reply pkg/domain/infra/abi/play.go:1537]: Added tests using your buildah idea.

File: pkg/api/handlers/libpod/kube.go [hunk 3]
```diff
@@ -89,21 +90,22 @@ func KubePlay(w http.ResponseWriter, r *http.Request) {
...
-		LogDriver:          logDriver,
-		LogOptions:         query.LogOptions,
+		Replace:            query.Replace,
+		ServiceContainer:   query.ServiceContainer,
 		StaticIPs:          staticIPs,
 		StaticMACs:         staticMACs,
-		IsRemote:           true,
+		UseLongAnnotations: query.NoTrunc,
+		Username:           username,
 		Userns:             query.Userns,
```
Comment (TomSweeneyRedHat) [inline pkg/api/handlers/libpod/kube.go:107]: I like the alpha ordering, but man, GitHub doesn't make it easy to review.

**Comment by TomSweeneyRedHat** [comment]
LGTM

File: pkg/api/handlers/libpod/kube.go (hunk 3 above)
Comment (vrothberg) [reply pkg/api/handlers/libpod/kube.go:107]: IMO such changes should be part of a separate commit.  That'll reduce the noise.

File: test/system/700-play.bats [hunk 4]
```diff
@@ -411,6 +411,23 @@ _EOF
...
 
+@test "podman kube play --replace external storage" {
+    TESTDIR=$PODMAN_TMPDIR/testdir
+    mkdir -p $TESTDIR
+    echo "$testYaml" | sed "s|TESTDIR|${TESTDIR}|g" > $PODMAN_TMPDIR/test.yaml
+    run_podman play kube $PODMAN_TMPDIR/test.yaml
+    # Force removal of container
+    run_podman rm --force -t0 test_pod-test
+    # Create container external using buildah with same name
+    buildah from --name test_pod-test $IMAGE
```
Comment (vrothberg) [inline test/system/700-play.bats:422]: This probably needs a `run buildah` and status checks.

**Review by vrothberg — APPROVED** [review]
Code LGTM

I'd prefer @edsantiago's blessing on the test.

File: test/system/700-play.bats (hunk 4 above)
Comment (edsantiago) [reply test/system/700-play.bats:422]: Not really. `buildah` here is much better than `run buildah`.

Reason: `run` is tricky. It hides exit status and output. In this case here, if plain `buildah` fails, BATS will immediately fail and will show all stdout/stderr. If you `run buildah`, though, you then need to add cumbersome `if $status blah` checks, and then will miss stdout/stderr if you check in the wrong order.

File: test/system/700-play.bats [hunk 5]
```diff
@@ -411,6 +411,23 @@ _EOF
...
 }
 
+@test "podman kube play --replace external storage" {
+    TESTDIR=$PODMAN_TMPDIR/testdir
+    mkdir -p $TESTDIR
+    echo "$testYaml" | sed "s|TESTDIR|${TESTDIR}|g" > $PODMAN_TMPDIR/test.yaml
+    run_podman play kube $PODMAN_TMPDIR/test.yaml
+    # Force removal of container
+    run_podman rm --force -t0 test_pod-test
+    # Create container external using buildah with same name
```
Comment (edsantiago) [inline test/system/700-play.bats:421]: However, I do find `container external` confusing. Should that be `external container`?

File: test/system/700-play.bats [hunk 6]
```diff
@@ -411,6 +411,23 @@ _EOF
     run_podman rmi -f userimage:latest
 }
 
+@test "podman kube play --replace external storage" {
```
Comment (edsantiago) [inline test/system/700-play.bats:414]: Also, as a general rule, when a test is super-confusing (like this one, WTF is it supposed to test) I like to add header comments like "Regression test for #12345; an external container, such as from buildah, caused a "container name is already in use" failure.

Comment (rhatdan) [reply test/system/700-play.bats:418]: Fixed.

**Review by vrothberg — APPROVED** [review]
/lgtm

**Comment by rhatdan** [comment]
Seems to work.
```
$ buildah from --name dan alpine
dan
$ $ podman run --name dan alpine echo hi
Error: creating container storage: the container name "dan" is already in use by d296206717c64cd5c2988f3aabcf89cf38e7a2010ea52f083c5fd63d7398fbeb. You have to remove that container to be able to reuse that name: that name is already in use
$ podman run --replace --name dan alpine echo hi
dan
hi
```

**Pull request merged**